EventFilter/DTRawToDigi/plugins/DTuROSRawToDigi.cc

```diff
@@ -0,0 +1,466 @@
+
+
+	  int wheelId, stationId, sectorId, slId,layerId, cellId;
+	  if ( ! mapping->readOutToGeometry(dduId, rosId, robId, tdcId, tdcChannel,
+					    wheelId, stationId, sectorId, slId, layerId, cellId)) {
+
+	    DTWireId detId = DTWireId(wheelId, stationId, sectorId, slId, layerId, cellId);
+	    int wire = detId.wire();
+
+	    DTDigi digi(wire, tdcTime, hitOrder[channelIndex.getCode()]);
+	    digis.insertDigi(detId.layerId(),digi);
+
+	  }
+
+	}
+        else if ( selector == 3 ) { // error word
+
+	  if (  debug_ ) edm::LogWarning("dturos_unpacker") << "Error word [" << std::dec << k << "] : " << std::hex << dataWord 
+							    << std::dec << " in slot " << slot << " in crate " << crate;
+
+	  int error  = ( dataWord >> 32 ) & 0x1FFFFFFF; // positions  32 -> 60
+	  rwords.seterror(error);
+
+	}
+
+
+	if ((dataWord & 0x1FFFFFFF) == 0x1FFFFFFF) continue;
+
+	if ( selector2 == 0 ) { // TDC word
+
+	  int tdcTime    = ( dataWord ) & 0x3FFF;       // positions   0 -> 13
+	  int tdcChannel = ( dataWord >> 14 ) & 0x1F;   // positions  14 -> 18
+	  int tdcId      = ( dataWord >> 19 ) & 0x3;    // positions  19 -> 20
+	  int link       = ( dataWord >> 21 ) & 0x7F;   // positions  21 -> 27
+
+
+	  if (tdcTime == 16383) continue;
+
+	  int dummy = 0;
+
+	  bool tenDDU = ! mapping->readOutToGeometry(779, 7, 1, 1, 1,
+						     dummy, dummy, dummy, 
+						     dummy, dummy, dummy);
+
+	  int dduId = theDDU(crate, slot, link, tenDDU);
+	  int rosId = theROS(crate, slot, link);
+	  int robId = theROB(crate, slot, link);
+
+
+	  DTROChainCoding channelIndex(dduId, rosId, robId, tdcId, tdcChannel);
+	  if (hitOrder.find(channelIndex.getCode()) == hitOrder.end()) hitOrder[channelIndex.getCode()] = 0;
+	  else hitOrder[channelIndex.getCode()]++;
+
+
+	  int wheelId, stationId, sectorId, slId,layerId, cellId;
+	  if ( ! mapping->readOutToGeometry(dduId, rosId, robId, tdcId, tdcChannel,
+					    wheelId, stationId, sectorId, slId, layerId, cellId)) {
+
+	    DTWireId detId = DTWireId(wheelId, stationId, sectorId, slId, layerId, cellId);
+	    int wire = detId.wire();
+
+	    DTDigi digi(wire, tdcTime, hitOrder[channelIndex.getCode()]);
+	    digis.insertDigi(detId.layerId(),digi);
+
+	  }
+
+	}
+        else if ( selector2 == 1 ) { // error word
+
+	  if (  debug_ ) edm::LogWarning("dturos_unpacker") << "Error word [" << std::dec << k << "] : " << std::hex << dataWord 
+							    << std::dec << " in slot " << slot << " in crate " << crate;
+
+	  int error  = ( dataWord ) & 0x1FFFFFFF;       // positions   0 -> 28
+	  rwords.seterror(error);
+
+	}
+
+      }
+
+    }
+
+
+    ++DTuROSiterator;
+    dataWord  = (*DTuROSiterator); // Trailer AMC
+
+
+    rwords.settrailer(dataWord);
+    fwords.setuROS(slot, rwords);
+
+  }  // end for-loop container content
+
+
+  return;
+}
+
+
+int DTuROSRawToDigi::theDDU(int crate, int slot, int link, bool tenDDU) {
+
+  int ros = theROS(crate,slot,link); 
+  
+  int ddu = 772;
```

Comment:
the "map" defined in this function can really never change in hardware?

Reply:
No, it cannot change in the hardware.
Thanks